### PR TITLE
MDEV-24927 & MDEV-25090: Deb: Use liburing-dev and libpmem-dev

### PR DESCRIPTION
**Pull request by ottok**
Updating the debian/control file will automatically update the dependencies
in all CI environments that directly read the debian/control file, such
as Salsa-CI and buildbot.mariadb.org to some degree.
(https://github.com/MariaDB/mariadb.org-tools/issues/43)

On Debian/Ubuntu releases that don't have liburing-dev available,
automatically downgrade to libaio-dev (just like libcurl4->3 is done).
This ensures the debian/control file is always up-to-date and works for
latest Debian and Ubuntu releases, while the backwards compatibility mods
are maintained in autobake-deb.sh separately, and can be dropped from there
once support for certains platforms end.

Also modify debian/rules to force a build without libaio. Use YES instead
of ON to make the flag more logical (=turning libaio check "off").

This commit does not modify the .travis.yml file, as Travis-CI does not
have new enough Ubuntu releases available yet. Also Travis-CI.org is
practically dead now as build times have been shrunk to near zero.

This commit does not update the external libmariadb or ColumnStore
CI pipelines, as those are maintained in different repositories.

The scope of this change is also Debian/Ubuntu only. No RPM or Windows or
Mac changes are included in this commit.

### CODE REVIEW

File: cmake/build_configurations/mysql_release.cmake [hunk 1]
```diff
@@ -149,7 +149,7 @@ IF(UNIX)
           RedHat/Fedora/Oracle Linux: yum install libaio-devel
           SuSE:                       zypper install libaio-devel
 
-          If you really do not want it, pass -DIGNORE_AIO_CHECK=ON to cmake.
+          If you really do not want it, pass -DIGNORE_AIO_CHECK=YES to cmake.
```
Comment (dr-m) [inline cmake/build_configurations/mysql_release.cmake:152]: Why this change? In `debian/rules` you are adding the value `-DIGNORE_AIO_CHECK=ON`. I believe that in `cmake`, `ON` and `YES` are interchangeable for `BOOL` parameters.

File: debian/autobake-deb.sh [hunk 2]
```diff
@@ -98,6 +98,13 @@ then
   sed 's/libcurl4/libcurl3/g' -i debian/control
 fi
 
+# From Debian Bullseye/Ubuntu Groovy, liburing replaces libaio
+if ! apt-cache madison liburing-dev | grep 'liburing-dev' >/dev/null 2>&1
+then
+  sed 's/liburing-dev/libaio-dev/g' -i debian/control
+  sed '/-DIGNORE_AIO_CHECK=YES/d' -i debian/rules
```
Comment (dr-m) [inline debian/autobake-deb.sh:105]: Should we remove more matches, such as `-DIGNORE_AIO_CHECK:BOOL=ON`?
I think that we would want to add `-DWITH_URING=ON` to `debian/rules` if the library is available. (This is waiting for MDEV-24883 to actually implement the build option.)

**Review by dr-m — CHANGES_REQUESTED** [review]
Thank you. I think that this is currently blocked by [MDEV-24883](https://jira.mariadb.org/browse/MDEV-24883), which in turn is waiting for an issue to be diagnosed on Ubuntu 20.04 (using its kernel and a custom built `liburing`).

My intention is that unlike with `libaio`, in the `cmake` invocation, one has to explicitly enable `-DWITH_URING=ON` (or `-DWITH_PMEM=ON`), and the configuration would fail if the dependency is not available.

Could you provide a similar solution for `libpmem-dev` and `-DWITH_PMEM=ON`? That option is already implemented in 10.5 but not enabled in our packages. I see that `libpmem-dev` is available in the stable Debian (Buster) but not earlier releases. I would like to enable the `libpmem` support in our 10.6 executables.

File: cmake/build_configurations/mysql_release.cmake (hunk 1 above)
Comment (ottok) [reply cmake/build_configurations/mysql_release.cmake:152]: It felt unintuitive to turn off some feature using "on".

File: debian/autobake-deb.sh (hunk 2 above)
Comment (ottok) [reply debian/autobake-deb.sh:105]: No, it only needs to match the exact lines in `debian/*` that exist there right now. It is not meant to be able to cope from arbitrary manual changes.

**Comment by ottok** [comment]
The build on http://buildbot.askmonty.org/buildbot/builders/kvm-deb-groovy-amd64/builds/646/steps/compile/logs/stdio did not seem to use uring in any way. The configure step also still says:
```
-- Looking for include file libaio.h
-- Looking for include file libaio.h - found
```

It is true that libaio-dev is installed, but the check should not run due the `debian/rules` changes.

File: cmake/build_configurations/mysql_release.cmake (hunk 1 above)
Comment (dr-m) [reply cmake/build_configurations/mysql_release.cmake:152]: Yes, I find this completely backwards too. I never liked the ‘add a dependency if the header files are present’ approach. The `libaio` dependency should have been ‘opt-in’ in the first place. OK, I agree with your reasoning.

**Comment by ottok** [comment]
## Testing/verification lacking

In https://salsa.debian.org/mariadb-team/mariadb-server/-/jobs/1498535/raw I can see:
```
-- Looking for include file libpmem.h
-- Looking for include file libpmem.h - found
```

There is nothing about uring (yet).

At this point in time I don't know how to test/verify if the binaries actually are using uring or pmem, but I assume that will later sort itself out as @dr-m develops on 10.6.

## 10.6 onwards only

**NOTE!** This PR is only for the 10.6 branch. The 10.5 branch is already stable and thus cannot (and should not) have any new dependencies added.

**Comment by ottok** [comment]
Updated 1st commit with `-DWITH_URING=ON` based on @dr-m comment on Zulip.

**Comment by ottok** [comment]
@kevgs If you are cherry-picking commits from this branch to other branches, please make sure you are using the latest version of this branch.

**Comment by ottok** [comment]
I rebased this on `bb-10.6-release` on @dr-m's request now.

E.g. http://buildbot.askmonty.org/buildbot/builders/kvm-deb-sid-amd64/builds/11823/steps/install/logs/stdio is now failing on:
```
Mar 11 19:01:29 debian-sid-amd64 mariadbd[2833]: io_uring_queue_init() failed with errno 12
Mar 11 19:01:29 debian-sid-amd64 mariadbd[2833]: 210311 19:01:29 [ERROR] mysqld got signal 11 ;
Mar 11 19:01:29 debian-sid-amd64 mariadbd[2833]: This could be because you hit a bug. It is also possible that this binary
Mar 11 19:01:29 debian-sid-amd64 mariadbd[2833]: or one of the libraries it was linked against is corrupt, improperly built,
Mar 11 19:01:29 debian-sid-amd64 mariadbd[2833]: or misconfigured. This error can also be caused by malfunctioning hardware.
Mar 11 19:01:29 debian-sid-amd64 mariadbd[2833]: To report this bug, see https://mariadb.com/kb/en/reporting-bugs
Mar 11 19:01:29 debian-sid-amd64 mariadbd[2833]: We will try our best to scrape up some info that will hopefully help
Mar 11 19:01:29 debian-sid-amd64 mariadbd[2833]: diagnose the problem, but since we have already crashed,
Mar 11 19:01:29 debian-sid-amd64 mariadbd[2833]: something is definitely wrong and this may fail.
Mar 11 19:01:29 debian-sid-amd64 mariadbd[2833]: Server version: 10.6.0-MariaDB-1:10.6.0+maria~sid
Mar 11 19:01:29 debian-sid-amd64 mariadbd[2833]: key_buffer_size=134217728
Mar 11 19:01:29 debian-sid-amd64 mariadbd[2833]: read_buffer_size=131072
Mar 11 19:01:29 debian-sid-amd64 mariadbd[2833]: max_used_connections=0
Mar 11 19:01:29 debian-sid-amd64 mariadbd[2833]: max_threads=153
Mar 11 19:01:29 debian-sid-amd64 mariadbd[2833]: thread_count=0
Mar 11 19:01:29 debian-sid-amd64 mariadbd[2833]: It is possible that mysqld could use up to
Mar 11 19:01:29 debian-sid-amd64 mariadbd[2833]: key_buffer_size + (read_buffer_size + sort_buffer_size)*max_threads = 467899 K  bytes of memory
Mar 11 19:01:29 debian-sid-amd64 mariadbd[2833]: Hope that's ok; if not, decrease some variables in the equation.
Mar 11 19:01:29 debian-sid-amd64 mariadbd[2833]: Thread pointer: 0x0
Mar 11 19:01:29 debian-sid-amd64 mariadbd[2833]: Attempting backtrace. You can use the following information to find out
Mar 11 19:01:29 debian-sid-amd64 mariadbd[2833]: where mysqld died. If you see no messages after this, something went
Mar 11 19:01:29 debian-sid-amd64 mariadbd[2833]: terribly wrong...
Mar 11 19:01:29 debian-sid-amd64 mariadbd[2833]: stack_bottom = 0x0 thread_stack 0x49000
Mar 11 19:01:29 debian-sid-amd64 mariadbd[2833]: /usr/sbin/mariadbd(my_print_stacktrace+0x2e)[0x56028f954b2e]
Mar 11 19:01:30 debian-sid-amd64 mariadbd[2833]: /usr/sbin/mariadbd(handle_fatal_signal+0x485)[0x56028f40f445]
Mar 11 19:01:30 debian-sid-amd64 mariadbd[2833]: sigaction.c:0(__restore_rt)[0x7efc26d38140]
Mar 11 19:01:30 debian-sid-amd64 mariadbd[2833]: ??:0(io_uring_get_sqe)[0x7efc26d1ab36]
Mar 11 19:01:31 debian-sid-amd64 mariadbd[2833]: /usr/sbin/mariadbd(_ZN5tpool16create_linux_aioEPNS_11thread_poolEi+0x174)[0x56028f8ee314]
Mar 11 19:01:31 debian-sid-amd64 mariadbd[2833]: /usr/sbin/mariadbd(+0xcb9bea)[0x56028f762bea]
Mar 11 19:01:31 debian-sid-amd64 mariadbd[2833]: include/liburing.h:192(tpool::create_linux_aio(tpool::thread_pool*, int))[0x56028f0d7599]
Mar 11 19:01:31 debian-sid-amd64 mariadbd[2833]: bits/unique_ptr.h:179(std::__uniq_ptr_impl<tpool::aio, std::default_delete<tpool::aio> >::reset(tpool::aio*))[0x56028f70b5b9]
Mar 11 19:01:31 debian-sid-amd64 mariadbd[2833]: srv/srv0start.cc:1185(srv_start(bool))[0x56028f411fee]
Mar 11 19:01:31 debian-sid-amd64 mariadbd[2833]: handler/ha_innodb.cc:3714(innodb_init(void*))[0x56028f219206]
Mar 11 19:01:31 debian-sid-amd64 mariadbd[2833]: sql/handler.cc:645(ha_initialize_handlerton(st_plugin_int*))[0x56028f21a345]
Mar 11 19:01:31 debian-sid-amd64 mariadbd[2833]: sql/sql_plugin.cc:1459(plugin_initialize(st_mem_root*, st_plugin_int*, int*, char**, bool))[0x56028f13e283]
Mar 11 19:01:31 debian-sid-amd64 mariadbd[2833]: sql/sql_plugin.cc:1751(plugin_init(int*, char**, int))[0x56028f144365]
Mar 11 19:01:32 debian-sid-amd64 mariadbd[2833]: csu/libc-start.c:308(__libc_start_main)[0x7efc26867d0a]
Mar 11 19:01:32 debian-sid-amd64 mariadbd[2833]: /usr/sbin/mariadbd(_start+0x2a)[0x56028f1395fa]
```

This at least proves the packaging part is correct and MariaDB is build with uring.


The builder with the full contents of this PR (with also the pmem commit) is failing at e.g. http://buildbot.askmonty.org/buildbot/builders/kvm-deb-sid-amd64/builds/11824/steps/compile/logs/stdio with:
```
Unmet build dependencies: libpmem-dev
```

This will start working once @dbart has updated the 10.6 builder to include package `libpem-dev` (tracked in https://jira.mariadb.org/browse/MDEV-25090).

**Comment by dr-m** [comment]
Thank you! I came across https://bugzilla.kernel.org/show_bug.cgi?id=211359 and was finally able to repeat the startup crash (`ENOMEM` followed by `SIGSEGV`) by setting the following:
```
sudo sh -
ulimit -l 0
gdb --args /usr/sbin/mysqld --datadir=/var/lib/mysql
run
```
With the default `ulimit -l` of 503282 bytes, the server would start up, but I would get another error on shutdown:
```
Thread 1 "mysqld" received signal SIGUSR1, User defined signal 1.
0x00007ffff765366f in __GI___poll (fds=fds@entry=0x555557593a60, nfds=nfds@entry=2, timeout=timeout@entry=-1) at ../sysdeps/unix/sysv/linux/poll.c:29
29	../sysdeps/unix/sysv/linux/poll.c: No such file or directory.
(gdb) c
Continuing.
2021-03-12  9:11:20 0 [Note] /usr/sbin/mysqld (initiated by: unknown): Normal shutdown
io_uring_wait_cqe() returned -4
2021-03-12  9:11:20 0 [Note] Event Scheduler: Purging the queue. 0 events
2021-03-12  9:11:20 0 [Note] InnoDB: FTS optimize thread exiting.
[Thread 0x7ffff44fd640 (LWP 9238) exited]
[Thread 0x7ffff44b2640 (LWP 9291) exited]

Thread 4 "mysqld" received signal SIGABRT, Aborted.
[Switching to Thread 0x7fffe1d1a640 (LWP 9226)]
__GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:49
49	../sysdeps/unix/sysv/linux/raise.c: No such file or directory.
(gdb) bt
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:49
#1  0x00007ffff756c864 in __GI_abort () at abort.c:79
#2  0x00005555563b70bf in (anonymous namespace)::aio_uring::thread_routine (aio=0x555557594850) at ./tpool/aio_liburing.cc:122
#3  0x00007ffff7957d84 in ?? () from /lib/x86_64-linux-gnu/libstdc++.so.6
#4  0x00007ffff7a76590 in start_thread (arg=0x7fffe1d1a640) at pthread_create.c:463
#5  0x00007ffff765f223 in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
```
Both the startup and shutdown crashes are something for @kevgs to address. My suggestion: Fall back to simulated AIO on `ENOMEM`, and allow the error on shutdown, just like we allow `EINTR` and `EINVAL` from the `io_getevents` system call.

@ottok I think that we need some magic that would specify `LimitMEMLOCK` in a systemd configuration file. I do not know what the minimal required value is, but we might go with the default 503282 that I observed on Ubuntu Groove. Can you implement this? Ideally, only if the `liburing` dependency is present.

**Comment by ottok** [comment]
> @ottok I think that we need some magic that would specify LimitMEMLOCK in a systemd configuration file. I do not know what the minimal required value is, but we might go with the default 503282 that I observed on Ubuntu Groove. Can you implement this? Ideally, only if the liburing dependency is present.

Sorry, I don't have capacity to do systemd changes (and the testing it requires) at the moment. Maintaining the Debian packaging part is quick an easy for me, so I can deliver it, but can't commit to systemd stuff at the moment.

**Comment by dr-m** [comment]
@ottok Fully understood. Your help with Debian packaging is greatly appreciated, and your changes seem to work fine.

I may have figured out where the systemd configuration file can be adjusted. We will track this further in [MDEV-24883](https://jira.mariadb.org/browse/MDEV-24883) and hopefully finish soon.

**Comment by ottok** [comment]
Rebased on latest bb-release-10.6. Builds fine on http://buildbot.askmonty.org/buildbot/grid?branch=ok-10.6-uring-pmem&category=main&category=experimental (apart from archs where libpmem-dev is still mssing).

Example:
```
-- Looking for include file libpmem.h
-- Looking for include file libpmem.h - found
...
cd /home/buildbot/buildbot/build/mariadb-10.6.0/builddir/mysys && /usr/bin/cc -DDBUG_TRACE -DHAVE_CONFIG_H -DHAVE_PCLMUL -DHAVE_PMEM -DHAVE_SSE42 -DHAVE_URING -D_FILE_OFFSET_BITS=64 -I/home/buildbot/buildbot/build/mariadb-10.6.0/wsrep-lib/include -I/home/buildbot/buildbot/build/mariadb-10.6.0/wsrep-lib/wsrep-API/v26 -I/home/buildbot/buildbot/build/mariadb-10.6.0/builddir/include -I/home/buildbot/buildbot/build/mariadb-10.6.0/include -I/home/buildbot/buildbot/build/mariadb-10.6.0/mysys -g -O2 -ffile-prefix-map=/home/buildbot/buildbot/build/mariadb-10.6.0=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -Wdate-time -D_FORTIFY_SOURCE=2 -pie -fPIC -fstack-protector --param=ssp-buffer-size=4 -O3 -g -static-libgcc -fno-omit-frame-pointer -fno-strict-aliasing  -Wno-uninitialized -fno-omit-frame-pointer -D_FORTIFY_SOURCE=2 -DDBUG_OFF -Wall -Wdeclaration-after-statement -Wextra -Wformat-security -Wno-format-truncation -Wno-init-self -Wno-nonnull-compare -Wno-unused-parameter -Wvla -Wwrite-strings   -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -o CMakeFiles/mysys.dir/mf_tempdir.c.o -c /home/buildbot/buildbot/build/mariadb-10.6.0/mysys/mf_tempdir.c
```
http://buildbot.askmonty.org/buildbot/builders/kvm-deb-sid-amd64/builds/11836/steps/compile/logs/stdio

**Comment by dr-m** [comment]
Thank you, I included the `liburing` part in 211e9b3e0a308094d8f71a7eb4e7fe9c64e3a911. I will keep this open until `WITH_PMEM` (`libpmem`) change is done. I will shortly submit my own version of that based on your changes.

**Comment by dr-m** [comment]
@ottok The remaining part of this is in 02fbb31c0b146cac668154019ab258d0421b5597 (also including some of my changes to make `-DWITH_PMEM=auto` the default). To me, it looks like some builders on buildbot.askmonty.org and buildbot.mariadb.org are missing some dependencies, but I did not notice any problems of this commit itself. What do you think?

**Comment by ottok** [comment]
I cleaned up this branch so that my two commits (that constitute this PR) are not interwoven with other people's commits. However this PR is still based on the bb-10.6-release branch so that the end result can build properly.

**Comment by ottok** [comment]
Rebased on 10.6 as requested by @dr-m. This PR includes additional cmake/pmem.cmake edits as suggested by Marko. Thanks!

**Comment by ottok** [comment]
Seems commits from this PR was already pushed on 10.6. It renders this PR moot, so I'll just abandon and close it.

**Comment by dr-m** [comment]
> Thank you, I included the `liburing` part in [211e9b3](https://github.com/MariaDB/server/commit/211e9b3e0a308094d8f71a7eb4e7fe9c64e3a911). I will keep this open until `WITH_PMEM` (`libpmem`) change is done. I will shortly submit my own version of that based on your changes.

I am sorry for the confusion. The `WITH_PMEM` (`libpmem-dev`) part is now in 10.6 as 418381bf99291b7faf24629426bb1c4a99cc943e and a similar change for other platforms in 7c5c6fa65c9f0ac8862baf8ab74ad69dd0beb4f8.
